### PR TITLE
Add `limit` info to `HashAgg` to speed up

### PR DESCRIPTION
**Pull request by CbcWestwolf**
tipb branch: https://github.com/CbcWestwolf/tipb/tree/limit-distinct

### CODE REVIEW

**Comment by TonsnakeLin** [comment]
If I understand correctly，limit is pushed to hashagg only when the current level select stmt has limit .
It‘s OK.

**Review by winoros — COMMENTED** [review]
Don't change anything in the plan build phase

**Review by winoros — COMMENTED** [review]
Actually, there are no changes for the TiDB side's aggregation. So the result is very likely to be a wrong one. Because we don't have very well concurrency control on the TiKV side. It's possible that TiKV returns more than one region's results and we need to do the limit again on the TiDB's side.

File: pkg/planner/core/logical_plan_builder.go [hunk 1]
```diff
@@ -1936,6 +1942,12 @@ func (b *PlanBuilder) buildDistinct(child LogicalPlan, length int) (*LogicalAggr
 	for i, col := range plan4Agg.schema.Columns {
 		col.RetType = plan4Agg.AggFuncs[i].RetTp
 	}
+	if limit != nil {
+		count, offset, err := extractLimitCountOffset(b.ctx, limit)
+		if err == nil && offset == 0 && count != 0 {
+			plan4Agg.limitCount = count
```
Comment (winoros) [inline pkg/planner/core/logical_plan_builder.go]: We can just set `limitCount` to `offset+count`. It's safe to do it.

**Pull request merged**